src/Test/Utilities/Desktop/ObjectReference.cs

```diff
@@ -0,0 +1,142 @@
+    /// to ensure the best possible patterns to avoid "gotchas" with these sorts of tests.
+    /// </summary>
+    public sealed class ObjectReference<T> where T : class
+    {
+        private T _strongReference;
+
+        /// <summary>
+        /// Tracks if <see cref="GetReference"/> was called, which means it's no longer safe to do lifetime assertions.
+        /// </summary>
+        private bool _strongReferenceRetrievedOutsideScopedCall;
+
+        private readonly WeakReference _weakReference;
+
+        public ObjectReference(T target)
+        {
+            if (target == null)
+            {
+                throw new ArgumentNullException(nameof(target));
+            }
+
+            _strongReference = target;
+            _weakReference = new WeakReference(target);
+        }
+
+        /// <summary>
+        /// Asserts that the underlying object has been released.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void AssertReleased()
+        {
+            ReleaseAndGarbageCollect();
+
+            Assert.False(_weakReference.IsAlive, "Reference should have been released but was not.");
+        }
+
+        /// <summary>
+        /// Asserts that the underlying object is still being held.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public void AssertHeld()
+        {
+            ReleaseAndGarbageCollect();
+
+            // Since we are asserting it's still held, if it is held we can just recover our strong reference again
+            _strongReference = (T)_weakReference.Target;
+            Assert.True(_strongReference != null, "Reference should still be held.");
+        }
+
+        // Ensure the mention of the field doesn't result in any local temporaries being created in the parent
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void ReleaseAndGarbageCollect()
+        {
+            if (_strongReferenceRetrievedOutsideScopedCall)
+            {
+                throw new InvalidOperationException($"The strong reference being held by the {nameof(ObjectReference<T>)} was retrieved via a call to {nameof(GetReference)}. Since the CLR might have cached a temporary somewhere in your stack, assertions can no longer be made about the correctness of lifetime.");
+            }
+
+            _strongReference = null;
+
+            // We'll loop 100 times, or until the weak reference disappears. When we're trying to assert that the
+            // object is released, once the weak reference goes away, we know we're good. But if we're trying to assert
+            // that the object is held, our only real option is to know to do it "enough" times; but if it goes away then
+            // we are definitely done.
+            for (var i = 0; i < 100 && _weakReference.IsAlive; i++)
+            {
+                GC.Collect();
```

Comment:
It looks like some final tweak I made means perhaps we have to restore that. :-(